frame/multi-account/src/lib.rs

```diff
@@ -0,0 +1,1506 @@
+		/// - `timepoint`: The timepoint (block number and transaction index) of the first approval
+		/// transaction for this dispatch.
+		/// - `call_hash`: The hash of the call to be executed.
+		///
+		/// # <weight>
+		/// - `O(S)`.
+		/// - Up to one balance-reserve or unreserve operation.
+		/// - One passthrough operation, one insert, both `O(S)` where `S` is the number of
+		///   signatories. `S` is capped by `MaxSignatories`, with weight being proportional.
+		/// - One encode & hash, both of complexity `O(S)`.
+		/// - One event.
+		/// - I/O: 1 read `O(S)`, one remove.
+		/// - Storage: removes one item.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(1_000_000)]
+		fn cancel(origin,
+			multi_account_id: T::AccountId,
+			timepoint: Timepoint<T::BlockNumber>,
+			call_hash: [u8; 32],
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let m = <Multisigs<T>>::get(&multi_account_id, call_hash)
+				.ok_or(Error::<T>::NotFound)?;
+			ensure!(m.when == timepoint, Error::<T>::WrongTimepoint);
+			ensure!(m.depositor == who, Error::<T>::NotOwner);
+
+			let _ = T::Currency::unreserve(&m.depositor, m.deposit);
+			<Multisigs<T>>::remove(&multi_account_id, call_hash);
+
+			Self::deposit_event(RawEvent::MultisigCancelled(who, timepoint, multi_account_id));
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Derive a multi account ID from the index.
+	pub fn multi_account_id(index: u64) -> T::AccountId {
+		let entropy = (b"modmua/multiacco", index).using_encoded(blake2_256);
+		T::AccountId::decode(&mut &entropy[..]).unwrap_or_default()
+	}
+
+	/// The current `Timepoint`.
+	pub fn timepoint() -> Timepoint<T::BlockNumber> {
+		Timepoint {
+			height: <system::Module<T>>::block_number(),
+			index: <system::Module<T>>::extrinsic_count(),
+		}
+	}
+
+	/// Check that signatories is sorted.
+	fn ensure_sorted(signatories: Vec<T::AccountId>) -> Result<Vec<T::AccountId>, DispatchError> {
+		let mut maybe_last = None;
+		for item in signatories.iter() {
+			if let Some(last) = maybe_last {
+				ensure!(last < item, Error::<T>::SignatoriesOutOfOrder);
+			}
+			maybe_last = Some(item);
+		}
+		Ok(signatories)
+	}
+
+	/// Check that signatories is sorted and doesn't contain sender, then insert sender.
+	fn ensure_sorted_and_insert(other_signatories: Vec<T::AccountId>, who: T::AccountId)
+		-> Result<Vec<T::AccountId>, DispatchError>
+	{
+		let mut signatories = other_signatories;
+		let mut maybe_last = None;
+		let mut index = 0;
+		for item in signatories.iter() {
+			if let Some(last) = maybe_last {
+				ensure!(last < item, Error::<T>::SignatoriesOutOfOrder);
+			}
+			if item <= &who {
+				ensure!(item != &who, Error::<T>::SenderInSignatories);
+				index += 1;
+			}
+			maybe_last = Some(item);
+		}
+		signatories.insert(index, who);
+		Ok(signatories)
+	}
+
+	/// Remove approvals from account IDs that are no longer in the set of signatories
```

Comment:
This looks very heavy. You are iterating over all approvals, and then doing a binary search for each of those.

Rather, you should make it so that the list of signatories cannot change if there are pending multisig operations.

Reply:
I like the approach you are suggesting, but it means that a former member can block it's removal by not removing a pending multisig operation.

In reality, there is an upper through `MaxSignatories`. So the upper boundary here is O(N * log(N)), where N is in reality most likely <= 100. 

An attack could be constructed by changing the signatories before the threshold is reached, so the approvals and signatories increase – but there is an economic cost to changing the signatories.

An alternative approach would be to clear all pending multisig operations on a signatory update, but finding all approvals at that stage is even more expensive. 

The logic here can be improved by first building a set and looking the signatories up on the set, but since N <= 100, it could be more expensive to build the set than to just do a binary search. 

Let me know what you think :)

Reply:
The problem is you are doing this with every single vote on a multi-sig, even when nothing has been changed. It certainly can be made more efficient. I am not sure what approach you want to take, but the one you have chosen here I don't think I would agree with.

I personally would not allow the multi-sig to change owners until all open multi-sig operations are closed. You can create a function where the multi-sig can close any operation on behalf of the original caller.

Reply:
@shawntabrizi that's a good alternative. 

I like the simplicity of what you are suggesting – it makes the whole logic much simpler and removes all the edge cases that need to be considered.

Just changed the code, documentation and tests accordingly – would appreciate your feedback.

Many thanks!